### PR TITLE
Calling natural_vector_size() on default Target should fail

### LINKED ISSUE #1104: Generator::natural_vector_size should assert when it is being called before the target is configured
Using Generator::natural_vector_size during construction of the generator object will result in a value that is probably incorrect, because it will use an un-configured Target.

### PR DESCRIPTION
**Pull request by steven-johnson**
Fixes #1104

### CODE REVIEW

**Comment by zvookin** [comment]
What I would like to happen is for natural_vector_size to return 0 or 1 for Targets that don't support vectorization and to have vectorize treat this as a nop. This allows writing code which vectorizes for systems where it is supported and does nothing for other ones.

**Comment by dsharletg** [comment]
+1 to @zvookin's idea.

**Comment by steven-johnson** [comment]
I like that idea, but (also) think that failing outright in the Target-is-default case is the right thing, since it ~always indicates a coding error.

**Comment by abadams** [comment]
Yeah, if natural_vector_size is used in a Generator member initializer,
you're going to have a bad time. Should be an explicit error. I also agree
that natural_vector size could usefully return 1 on some platforms, but
that seems like a separate issue.

File: src/Target.h [hunk 1]
```diff
@@ -237,6 +237,8 @@ struct Target {
     /** Given a data type, return an estimate of the "natural" vector size
      * for that data type when compiling for this Target. */
     int natural_vector_size(Halide::Type t) const {
+        user_assert(*this != Target()) << "natural_vector_size cannot be used on an uninitialized Target.\n";
```
Comment (zvookin) [inline src/Target.h]: The wording "uninitialized" is incorrect here. Given that we use the word "unknown" in OSUnknown and ArchUnknown, I would say "unknown" target. "Default" is also acceptable.

**Comment by dsharletg** [comment]
I was picturing fixing #1104 by adding some code to the GeneratorParam class. I suppose I hadn't thought that through at the time, but now that I think about it more, it seems like that issue affects any GeneratorParam, not just target.

File: src/Target.h (hunk 1 above)
Comment (steven-johnson) [inline src/Target.h]: Done

**Comment by steven-johnson** [comment]
So this is morphing into three issues:
-- Targets with Unknown fields should fail hard.
-- Targets that are well-defined but don't support vectorization should return 0 or 1 for quiet failure.
-- GeneratorParams that aren't initialized should fail hard.

I think that this patch solves #1104 as written, and we should open new issues for the others.

**Pull request merged**